File: services/agent/src/utils/knowledge-base.ts
/**
 * Knowledge Base Service Utilities
 * Handles knowledge base search and context retrieval
 */

import { fetchWithTimeout } from './fetch-with-timeout.js'
import { createLogger } from '@syntera/shared/logger/index.js'

const logger = createLogger('agent-service:knowledge-base')

const KNOWLEDGE_BASE_SERVICE_URL = process.env.KNOWLEDGE_BASE_SERVICE_URL
if (!KNOWLEDGE_BASE_SERVICE_URL) {
  throw new Error('KNOWLEDGE_BASE_SERVICE_URL environment variable is required')
}
const DEFAULT_TIMEOUT = 10000 // 10 seconds

export interface KnowledgeBaseSearchOptions {
  query: string
  companyId: string
  agentId: string
  topK?: number
  maxResults?: number
}

export interface KnowledgeBaseResult {
  text: string
  score?: number
}

/**
 * Search knowledge base and return formatted context string
 * 
 * Queries the Knowledge Base Service for relevant documents using vector search.
 * Returns a formatted context string that can be injected into AI prompts.
 * 
 * @param options - Search options
 * @param options.query - Search query text
 * @param options.companyId - Company ID to filter documents
 * @param options.agentId - Agent ID to filter documents
 * @param options.topK - Number of top results to retrieve (default: 5)
 * @param options.maxResults - Maximum results to include in context (default: 3)
 * @returns Formatted context string with document excerpts, or undefined if no results
 */
export async function searchKnowledgeBase(
  options: KnowledgeBaseSearchOptions
): Promise<string | undefined> {
  const { query, companyId, agentId, topK = 5, maxResults = 3 } = options

  try {
    const searchResponse = await fetchWithTimeout(
      `${KNOWLEDGE_BASE_SERVICE_URL}/api/documents/search`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          query,
          companyId,
          agentId,
          topK,
        }),
      },
      DEFAULT_TIMEOUT
    )

    if (!searchResponse.ok) {
      return undefined
    }

    const searchData = (await searchResponse.json()) as {
      results?: Array<{ metadata?: { [key: string]: unknown }; score?: number }>
    }

    if (!searchData.results || searchData.results.length === 0) {
      return undefined
    }

    // Format knowledge base results as context
    const contextTexts = searchData.results
      .slice(0, maxResults)
      .map((result) => {
        const text = result.metadata?.text
        return typeof text === 'string' ? text : ''
      })
      .filter((text: string) => text.length > 0)

    if (contextTexts.length === 0) {
      return undefined
    }

    return contextTexts.join('\n\n---\n\n')
  } catch (error) {
    logger.warn('Failed to retrieve knowledge base context', {
      error: error instanceof Error ? error.message : String(error),
      query: query.substring(0, 50),
    })
    return undefined
  }
}

File: services/voice-agent/src/knowledge_base.py
"""Knowledge Base integration for voice agent"""

import asyncio
import aiohttp
from typing import Optional
from config import settings
from utils.logger import setup_logger

logger = setup_logger(__name__)


async def get_knowledge_base_context(
    query: str,
    company_id: str,
    agent_id: str,
    top_k: int = 5
) -> Optional[str]:
    """
    Retrieve knowledge base context for a query
    
    Args:
        query: The search query (user message or conversation context)
        company_id: Company ID for filtering documents
        agent_id: Agent ID for filtering documents
        top_k: Number of results to return
        
    Returns:
        Combined knowledge base context as string, or None if no results
    """
    try:
        kb_url = settings.knowledge_base_service_url
        search_url = f"{kb_url}/api/documents/search"
        
        async with aiohttp.ClientSession() as session:
            async with session.post(
                search_url,
                json={
                    "query": query,
                    "companyId": company_id,
                    "agentId": agent_id,
                    "topK": top_k,
                },
                headers={"Content-Type": "application/json"},
                timeout=aiohttp.ClientTimeout(total=5),  # 5 second timeout
            ) as response:
                if response.status == 200:
                    data = await response.json()
                    results = data.get("results", [])
                    
                    if results:
                        # Extract text from results
                        context_parts = []
                        for result in results:
                            metadata = result.get("metadata", {})
                            text = metadata.get("text", "")
                            if text:
                                context_parts.append(text)
                        
                        if context_parts:
                            context = "\n\n".join(context_parts)
                            return context
                
                logger.debug("No knowledge base results found", {
                    "agent_id": agent_id,
                    "status": response.status,
                })
                return None
                
    except (asyncio.TimeoutError, aiohttp.ClientError) as e:
        logger.warn("Knowledge base search failed", {
            "agent_id": agent_id,
            "error": str(e),
        })
        return None
    except Exception as e:
        logger.warn("Failed to retrieve knowledge base context", {
            "error": str(e),
            "agent_id": agent_id,
        })
        return None


def enhance_system_prompt_with_kb(
    system_prompt: str,
    kb_context: Optional[str]
) -> str:
    """
    Enhance system prompt with knowledge base context
    
    Args:
        system_prompt: Original system prompt
        kb_context: Knowledge base context to add
        
    Returns:
        Enhanced system prompt with KB context
    """
    if not kb_context:
        return system_prompt
    
    enhanced = system_prompt
    
    # Add KB context if not already present
    if "knowledge base" not in enhanced.lower() and "relevant context" not in enhanced.lower():
        enhanced += f"\n\nRelevant context from knowledge base:\n{kb_context}"
    else:
        # If KB context section exists, append to it
        enhanced += f"\n\n{kb_context}"
    
    return enhanced

